File: sdks/typescript/src/opik/client/payloadTruncation.ts
import { logger } from "@/utils/logger";

/**
 * Per-object payload-size enforcement for spans and traces (parity with the
 * Python SDK, OPIK-7335).
 *
 * A span or trace whose `input`/`output` fields are very large - e.g. an entire
 * retrieval result set logged inline - inflates into a multi-GB structure on the
 * backend and can destabilise ingestion. This enforces a per-object size limit
 * with a simple, predictable two-pass rule:
 *   1. truncate any field that on its own exceeds the limit (common case: one
 *      giant field; small siblings kept);
 *   2. if input+output together are still over the limit, truncate the remaining
 *      truncatable field too.
 *
 * `metadata` is deliberately never truncated (it holds small structured fields
 * consumers rely on) and is excluded from the whole-object measurement, so a large
 * metadata can't trigger truncation of input/output. The oversized field is
 * replaced with a compact marker; a warning is logged. Non-mutating: returns a
 * shallow copy with the markers applied.
 */

// Only input/output are truncatable; metadata is exempt (see the note above).
const TRUNCATABLE_FIELDS = ["input", "output"] as const;
type TruncatableField = (typeof TRUNCATABLE_FIELDS)[number];

// Minimal shape the truncation works on - satisfied by SavedSpan / SavedTrace (create)
// and their update payloads. Constraining to this (rather than Record<string, unknown>)
// keeps interface types like SavedSpan/SavedTrace assignable.
type PayloadLike = { input?: unknown; output?: unknown; metadata?: unknown };

// What kind of object is being truncated - used only for the log message.
type PayloadKind = "span" | "trace";

const BYTES_PER_MB = 1024 * 1024;

interface TruncationMarker {
  // snake_case: this is wire data stored on the object, matching the backend / Python SDK marker.
  opik_truncated: true;
  reason: string;
}

const truncationMarker = (sizeMb: number): TruncationMarker => ({
  opik_truncated: true,
  reason: `<omitted_due_to_size_${Math.round(sizeMb)}MB_error_code_413_400>`,
});

// V8 caps a single string at ~512 MiB; JSON.stringify throws RangeError when its output would
// exceed that. Such a field is at least this large, so we report this lower bound (rather than
// 0) to force truncation of the very multi-GB payloads this guard exists to catch.
const MAX_SERIALIZABLE_MB = 512;

// Serialized size of a value in MB - what it would weigh on the wire. A RangeError means the
// value is too large to serialize (see above) and must be truncated. Any other failure
// (circular reference, BigInt, ...) is genuinely unmeasurable, so return 0 and leave it
// untouched. Truncation must never throw and break span/trace creation.
const fieldSizeMb = (value: unknown): number => {
  try {
    const json = JSON.stringify(value);
    return json ? Buffer.byteLength(json, "utf8") / BYTES_PER_MB : 0;
  } catch (error) {
    return error instanceof RangeError ? MAX_SERIALIZABLE_MB : 0;
  }
};

/**
 * Return a shallow copy of `payload` with oversized fields replaced by a truncation
 * marker (or `payload` unchanged), plus the list of fields that were truncated.
 */
export const truncatePayloadFields = <T extends PayloadLike>(
  payload: T,
  maxSizeMb: number,
): { result: T; truncated: TruncatableField[] } => {
  const sizes = {} as Record<TruncatableField, number>;
  for (const field of TRUNCATABLE_FIELDS) {
    if (payload[field] != null) {
      sizes[field] = fieldSizeMb(payload[field]);
    }
  }
  if (Object.keys(sizes).length === 0) {
    return { result: payload, truncated: [] };
  }

  const overrides: Partial<Record<TruncatableField, TruncationMarker>> = {};

  // Pass 1 - fields individually over the limit (the common "one giant field").
  for (const field of TRUNCATABLE_FIELDS) {
    if (sizes[field] !== undefined && sizes[field] > maxSizeMb) {
      overrides[field] = truncationMarker(sizes[field]);
    }
  }

  // Pass 2 - hard per-object cap: if input+output are still over as a whole,
  // truncate the remaining truncatable field too. One measurement, no loop.
  // metadata is excluded so it can't drag the object over and cut small siblings.
  if (fieldSizeMb({ ...payload, ...overrides, metadata: undefined }) > maxSizeMb) {
    for (const field of TRUNCATABLE_FIELDS) {
      if (sizes[field] !== undefined && overrides[field] === undefined) {
        overrides[field] = truncationMarker(sizes[field]);
      }
    }
  }

  const truncated = Object.keys(overrides) as TruncatableField[];
  if (truncated.length === 0) {
    return { result: payload, truncated: [] };
  }
  return { result: { ...payload, ...overrides } as T, truncated };
};

/**
 * Truncate a span/trace payload (create or update) if its fields exceed the
 * per-object limit, logging a warning. Returns a shallow copy (or the input
 * unchanged). A `maxSizeMb <= 0` disables the check. `kind`/`id` are used only
 * for the log message.
 */
export const truncatePayloadIfNeeded = <T extends PayloadLike>(
  payload: T,
  maxSizeMb: number,
  kind: PayloadKind = "span",
  id?: string,
): T => {
  if (!maxSizeMb || maxSizeMb <= 0) {
    return payload;
  }
  const { result, truncated } = truncatePayloadFields(payload, maxSizeMb);
  if (truncated.length > 0) {
    const label = kind.charAt(0).toUpperCase() + kind.slice(1);
    logger.warn(
      `${label} '${id ?? "unknown"}' exceeded the payload size limit of ${maxSizeMb} MB; ` +
        `truncated field(s): ${truncated.join(", ")}. ` +
        `Log large payloads as attachments to avoid truncation.`,
    );
  }
  return result;
};

File: sdks/python/src/opik/message_processing/payload_truncation.py
"""Per span/trace payload-size enforcement.

A span or trace whose ``input``/``output`` fields are very large - e.g. an entire
retrieval result set logged inline - inflates into a multi-GB structure on the
backend and can destabilise ingestion. This module enforces a **per-object** size
limit with a simple, predictable two-pass rule:

1. any field that on its own exceeds the limit is replaced with a truncation
   marker (the common case: one giant field, small siblings kept);
2. if the object as a whole is still over the limit, the remaining truncatable
   fields are replaced too - so the object is guaranteed to end up under the limit.

No field sorting and no per-step re-measure loop (at most one whole-object
measurement). Applied in the background message processor, right before the
create/update request is sent - after attachments have been extracted/uploaded
separately, so only content that would be sent inline is measured and truncated.
A warning is logged whenever anything is truncated.

``metadata`` is deliberately never truncated (it holds small structured fields such
as ``thread_id``/``model`` that the backend relies on) and is excluded from the
whole-object measurement, so a large ``metadata`` cannot trigger truncation of
``input``/``output``.
"""

import logging
from typing import Any, Callable, Dict, List, Optional, TypeVar

from .batching import sequence_splitter
from ..rest_api.types import span_write, trace_write

LOGGER = logging.getLogger(__name__)

# A span or trace write model - both carry the same truncatable input/output fields.
WriteT = TypeVar("WriteT", span_write.SpanWrite, trace_write.TraceWrite)

# Fields on a span/trace that can carry large user-provided payloads. Deliberately
# only `input`/`output` - `metadata` is left untouched (see the module docstring).
_TRUNCATABLE_FIELDS = ("input", "output")


def _truncation_marker(size_mb: float) -> Dict[str, Any]:
    # Compact marker left in place of an oversized field. `opik_truncated` is the
    # machine-detectable flag; `reason` is a short tag with the original size and the
    # server error codes (413/400) the field would otherwise trigger.
    return {
        "opik_truncated": True,
        "reason": f"<omitted_due_to_size_{round(size_mb)}MB_error_code_413_400>",
    }


def _log_truncation(
    kind: str, obj_id: Optional[str], max_size_mb: float, truncated_fields: List[str]
) -> None:
    LOGGER.warning(
        "%s '%s' exceeded the per-%s size limit of %s MB; truncated field(s): %s. "
        "Log large payloads as attachments to avoid truncation.",
        kind.capitalize(),
        obj_id,
        kind,
        max_size_mb,
        ", ".join(truncated_fields),
    )


def _field_sizes_mb(get_field: Callable[[str], Any]) -> Dict[str, float]:
    sizes: Dict[str, float] = {}
    for name in _TRUNCATABLE_FIELDS:
        value = get_field(name)
        if value is not None:
            sizes[name] = sequence_splitter.get_payload_size_MB(value)
    return sizes


def _plan_truncation(
    field_sizes_mb: Dict[str, float],
    max_size_mb: float,
    measure_whole: Callable[[Dict[str, Any]], float],
) -> Dict[str, Any]:
    """Decide which fields to replace with a marker so the object ends up <= the limit.

    Pass 1 truncates any field that alone exceeds the limit. Pass 2 (the hard
    per-object cap) truncates the rest only if the object is still over as a whole.
    Returns ``{field_name: marker}`` (empty if nothing needs truncating).
    """
    updates: Dict[str, Any] = {}

    # Pass 1 - fields individually over the limit (the common "one giant field").
    for name, size_mb in field_sizes_mb.items():
        if size_mb > max_size_mb:
            updates[name] = _truncation_marker(size_mb)

    # Pass 2 - hard per-object cap: if it is still over as a whole, truncate the
    # remaining truncatable fields too. One measurement, no loop.
    if measure_whole(updates) > max_size_mb:
        for name, size_mb in field_sizes_mb.items():
            if name not in updates:
                updates[name] = _truncation_marker(size_mb)

    return updates


def truncate_write_if_needed(
    obj: WriteT, max_size_mb: float, kind: str = "span"
) -> WriteT:
    """Return a copy of ``obj`` (a span or trace write) with oversized fields truncated."""
    # A non-positive limit disables the check (parity with the TS SDK). Guarding here also
    # avoids the degenerate case where a 0/negative limit would mark every field oversized.
    if max_size_mb <= 0:
        return obj
    field_sizes = _field_sizes_mb(lambda n: getattr(obj, n, None))
    if not field_sizes:
        return obj

    def measure_whole(overrides: Dict[str, Any]) -> float:
        # Exclude metadata: it is never truncated, so it must not drag the object
        # "over" the cap and trigger pointless truncation of small input/output.
        candidate = obj.model_copy(update={"metadata": None, **overrides})
        return sequence_splitter.get_payload_size_MB(candidate)

    updates = _plan_truncation(field_sizes, max_size_mb, measure_whole)
    if not updates:
        return obj

    # The write models are frozen; model_copy(update=...) builds a new instance
    # without re-validating, so the marker dict is accepted on the JSON fields.
    result = obj.model_copy(update=updates)
    _log_truncation(kind, getattr(obj, "id", None), max_size_mb, list(updates))
    return result


def truncate_writes(
    objs: List[WriteT], max_size_mb: float, kind: str = "span"
) -> List[WriteT]:
    """Truncate every span/trace in a batch that exceeds the per-object limit."""
    return [truncate_write_if_needed(obj, max_size_mb, kind) for obj in objs]


def truncate_kwargs_if_needed(
    kwargs: Dict[str, Any], max_size_mb: float, kind: str = "span"
) -> None:
    """Truncate oversized fields on a single create/update payload dict, in place.

    Used on the non-batched create path (``use_batching=False``) and on the
    update path, so an oversized ``output``/``input`` sent via an update
    (e.g. ``span.end(output=...)`` after the create was already flushed) is capped
    the same way as on create. Applies to both spans and traces (``kind``).
    """
    # A non-positive limit disables the check (parity with the TS SDK).
    if max_size_mb <= 0:
        return
    field_sizes = _field_sizes_mb(kwargs.get)
    if not field_sizes:
        return

    def measure_whole(overrides: Dict[str, Any]) -> float:
        # Exclude metadata (never truncated) from the whole-object measurement.
        return sequence_splitter.get_payload_size_MB(
            {**kwargs, **overrides, "metadata": None}
        )

    updates = _plan_truncation(field_sizes, max_size_mb, measure_whole)
    if not updates:
        return

    kwargs.update(updates)
    _log_truncation(kind, kwargs.get("id"), max_size_mb, list(updates))
